File: flesch.java
//Colton Steinbeck//
//Test comment for the file transfer on Anvil//
import java.lang.*;
import java.util. *;
import java.io.*;
import java.Scanner;

public class flesch
{    

    public static void main(Strings[] args) 
    {
        //Takes the users command line input
    System.out.println("Please enter specific file locations for translations: ");
    Scanner tScan = new Scanner(System.in);
    String translation = tScan.nextLine();

    //Takes the file path puts these words into the array//
    File translation = new File();
    Scanner scan = new Scanner(File(translation));
     
    ArrayList<String> vals= new ArrayList<String>();
    
        while(scan.hasNext())
        {
            String words= scan.next();
            vals.add(words);
        
        }


        
        //Pulls the Dale-Chall words into storage 
        Scanner dc = new Scanner(System.in("/pub/pounds/CSC330/dalechall"));
        Map<String, Object> dcStorage = new HashMap<String, Object>();
        
        while(dc.hasNextLine()) 
        {
            String word = dc.nextLine();
            word=word.toLowerCase();
            dcStorage.put(word,null);
        }

        int words = 0;
        int syllables = 0;
        int sentence = 0;


        //Goes through the whole array of the Bible (Main Arra)
        for(int i=0 ;i<=vals.size(); i++)
        {
            String token = vals.get(i);
            token = token.getLowerCase(); 
            
            //Check for punctuation
           for(int j=0; j<=token.length();j++)
           {
               char punc = token.charAt(j);
               if(punc == '.' || punc =='!' || punc == '?' || punc == ':' || punc == ';' )
               {
                    sentence++;
               }
               //Determines if the 
               if (!isInteger(token))
               {
                   word++;
                   syllables = syllables + numSyllables(token);
        
               }
        }


        







        }
        System.out.println("Word count: " + word);
		System.out.println("Sentence count: " + sentence);
		System.out.println("Syllable count: " + syllables);
		System.out.println("Difficult Word count: " + diffWord);

		double alpha = ((double)syllables / (double)word);
		double beta = ((double)word / (double)sentence);
		double gamma = ((double)diffWord / (double)word);

		double flesch = 206.835 - (alpha * 84.6) - (beta * 1.015);
		double grade = (alpha * 11.8) + (beta * 0.39) - 15.59;
		double readability;
        if (gamma > 0.05)
        {
            readability = ((gamma * 100.0) * 0.1579) + (beta * 0.0496) + 3.6365;
        }
        else
        {
            readability = ((gamma * 100.0) * 0.1579) + (beta * 0.0496);
        }

		System.out.println("The Flesch Readability index is: " + flesch);
		System.out.println("The Flesh-Kincaid Grade Level index is: " + grade);
		System.out.println("The Dale-Chall Readability Score is: " + readability);




    }
    //Checks if it is a vowel
    private static boolean isVowel(char letter)
    {
        return letter == 'a' || letter == 'e' || 
               letter == 'i' || letter == 'o' ||
               letter == 'u' || letter == 'y';
    }

    //Counts the number of syllables in the 
    private static int syllableCount(String word)
    {
       int syllableCount=0;
        char state= "x";
        // 1. Adjacent vowels to count as syllable  
        for(auto state : word)
        {
            switch(state)
            {
                case 'x':
                break;
                
                
                case 'v':
                {
                    if(!isVowel(x)) //Makes sure that count will not increment if there is a vowel next to it
                    {
                        syllableCount++;
                    }
                    break;
            }
            }   
            state = isVowel(x)? 'v':'x';
        }
         if(state=='v')
         {
             syllableCount++;
         }


            
            //2. Special Case with e at the end of the word
            if(letter == 'e' && i= word.length()-1)
            {
                i++;
            }
           

          else
          {
              
          } 
        


    }


    private static boolean isNumber(int value)
    {

        try 
        {
        Integer.parseInt(value);
        return true;    
        } 

        catch (Exception e) 
        {
            return false;
        }
    }
}

File: flesch.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstring>
#include <unordered_map>
#include <vector>
#include <cmath>
using namespace std;

//Checks if the character is a vowel
bool IsVowel (char letter)
{
if(letter == 'a' || letter == 'e' || letter == 'i' || letter == 'o' || letter == 'u' || letter =='y')
{
    return true;
}
return false;
} 


//Method to count the syllables in each word
int syllableCount( std::string word ) 
{
    int count = 0;
    char state = 'x';
    for (auto x : word ) 
    {
        switch(state) 
        {
        case 'x': 
        {
            break;
        }
        case 'v': 
        {
            if (!IsVowel(x))
            {
                count++;
            };
            break;
        }
        }
        state = IsVowel(x)?'v':'x';
    }
            if (state == 'v') 
            {
                count++;
            }
            return count;
  }

//Checks if the value is an integer
bool isNum(string token)
{
    for (int i = 0; i < token.length(); i++)
    {
        if (isdigit(token[i]) == false)
        { 
            return false;
        } 
    }
    return true; 
}

//Main Function
int main()
{

    //Command to input file location
    cout<<"Please enter specific file locations for translations:  "<<endl;
    string input;
    cin>>input;



//////////Storage
    //Checks if the file has opened
    ifstream BibleFile(input);
    BibleFile.open(input);
        if(BibleFile.fail())
        {
            cout<<"File cannot be accessed, please try again"<<endl;
        }
            //Imports the Dale-Chall List of Dif
    ifstream dalechall("/pub/pounds/CSC330/dalechall/wordlist1995.txt");
    	if (dalechall.is_open())
        {
		    cout << "The Dale-Chall list opened correctly." << endl;
        }
	    else 
        {
            cout << "Unable to open Dale-Chall list of words." << endl; 
        }


           //Creation of Storage for the Translations
            vector<string> vals;
            string tmp;
            
        while(getline(BibleFile,tmp))
        {
            stringstream s1(tmp);
            while(getline(s1,tmp,' '))
            {
                vals.push_back(tmp);
            }
        }
    unordered_map <string,int> dcStorage; //Storage for the Dale-Chall List
    string temp;
    int iteration=0;
    for(dalechall >> temp)
    {
        iteration++;
        for(int i=0;i<temp.size();i++)  
        {
            if(ispunct(temp[i]))
            {
                temp.erase(i--,1); //Removes the value if happens to be puncuation
            }
            temp[i]=tolower(temp[i]);
        } 
        dcStorage.insert(make_pair(temp, iteration)); //Inserts Map Pairing for value and placement
    }

    int word=0;
    int syllables=0;
    int sentence=0;
    int difficultWord=0;

    for(int i=0;i<vals.size();i++)
    {
        string instance = vals[i];
        
        if(!isNum(instance))
        {
            word++;
        } //Checks if the parameter passed in is a number

        for(int j=0;j<instance.length();i++)
        {
            char punc= instance[j];
            if(punc == '.' || punc =='?' || punc =='!' || punc ==';'|| punc ==':')
            {
                sentence++;
                instance = instance.substr(0,instance.length()-1); //Decreases the amount of items in the vector
            }
            syllable= syllable+ syllableCount(instance);
        }
    }

	cout << "Word count: " << word << endl;	
	cout << "Sentence count: " << sentence << endl;
	cout << "Syllable count: " << syllable << endl;
	cout << "Difficult Word count: " << difficultWord << endl;

	double alpha = ((double)syllable / (double)word);
	double beta = ((double)word / (double)sentence);
	double gamma = ((double)difficultWord / (double)word) * 100;

	double index = (206.835 - (alpha * 84.6) - (beta * 1.015));
	double grade = ((alpha * 11.8) + (beta * 0.39) - 15.59);

	cout << "The Flesch index score is: " << round(index) << endl;
	cout << "The Flesch Kincaid Grade Level index is: " << grade << endl;

	double readability;

	if (gamma > 5.0)
		readability = (gamma * 0.1579) + (beta * 0.0496) + 3.6365;
	else
		readability = (gamma * 0.1579) + (beta * 0.0496);

	cout << "The Dale-Chall Readability score is: " << readability << endl; 

	BibleFile.close();	

	return 0;
}















